lib/plugins/aws/package/compile/events/apiGateway/lib/requestValidator.js

```diff
@@ -0,0 +1,213 @@
+              this.provider.naming.getModelLogicalId(resourceName)
+            );
+
+            Object.assign(
+              this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
+              {
+                [modelLogicalId]: {
+                  Type: 'AWS::ApiGateway::Model',
+                  Properties: {
+                    RestApiId: this.provider.getApiGatewayRestApiId(),
+                    ContentType: contentType,
+                    Schema: definition,
+                  },
+                },
+
+                [validatorLogicalId]: {
+                  Type: 'AWS::ApiGateway::RequestValidator',
+                  Properties: {
+                    RestApiId: this.provider.getApiGatewayRestApiId(),
+                    ValidateRequestBody: true,
+                    ValidateRequestParameters: true,
+                  },
+                },
+              }
+            );
+
+            if (modelName) {
+              this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
+                modelLogicalId
+              ].Properties.Name = modelName;
+              this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
+                validatorLogicalId
+              ].Properties.Name = validatorName;
+            }
+
+            if (description) {
+              this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
+                modelLogicalId
+              ].Properties.Description = description;
+            }
+          } else if (apiGateway.requestSchemas && apiGateway.requestSchemas[schemaConfig]) {
+            const models = this.createProviderModel(
+              schemaConfig,
+              apiGateway.requestSchemas[schemaConfig]
+            );
+            modelLogicalId = models.modelLogicalId;
+            validatorLogicalId = models.validatorLogicalId;
+          }
+
+          template.Properties.RequestValidatorId = { Ref: validatorLogicalId };
+          template.Properties.RequestModels = template.Properties.RequestModels || {};
+          template.Properties.RequestModels[contentType] = { Ref: modelLogicalId };
+        }
+      }
+    });
+
+    return BbPromise.resolve();
+  },
+
+  createProviderModel(schemaId, schemaConfig) {
+    let modelName;
+    let validatorName;
+    let description;
+    let definition;
+
+    // If schema is not defined this will try to map resourceDefinition as the schema
+    if (!schemaConfig.schema) {
+      definition = schemaConfig;
+    } else {
+      definition = schemaConfig.schema;
+    }
+
+    const modelLogicalId = this.provider.naming.getModelLogicalId(schemaId);
+
+    const validatorLogicalId = this.provider.naming.getValidatorLogicalId(modelLogicalId);
+
+    if (schemaConfig.name) {
+      modelName = schemaConfig.name;
+      validatorName = `${modelName}Validator`;
+    }
+
+    if (schemaConfig.description) {
+      description = schemaConfig.description;
+    }
```

Comment:
I believe that's not the case - if the schema is defined as `application/json: '${file(dummySchema.json)}'` it is also going to be an object but without `schema` property

Reply:
My point was that by AJV schema, we allow two type of values here for `schemaConfig` value, it's _string_ and _object_ -> https://github.com/serverless/serverless/blob/0ba10467bb6ae9162d8fd9d1f4665a8bc0e264cd/lib/plugins/aws/package/compile/events/apiGateway/index.js#L121

And if that' s valid, it would be nice to clearly separate handling of both cases. I don't see it being done in this logic.

Reply:
Sorry, I believe I've misunderstood it a bit in the beginning - I've refactored the code to separate different branches:
- `schemaConfig` is a string and references definition from provider
- `schemaConfig` is a string that does not reference definition from provider
- `schemaConfig` is an object with implicit schema definition (e.g `'application/json': '${file(dummySchema.json)}')
- `schemaConfig` is an object with explicit schema definition

We cannot cleanly separate string vs object because one string-based version falls under the same processing flow as object-based one but I believe it's now a bit more clear in the code.

Reply:
Thanks @pgrzesik I think I also got a bit lost, as here as you pointed we will deal just with object notation.
Great job on clarifying that in upper part of the code.